app/request/context.go

```diff
@@ -39,8 +39,11 @@ func NewContext(ctx context.Context, requestId, ipAddress, path, userAgent, acce
 }
 
 func EmptyContext() *Context {
+	logger, _ := mlog.NewLogger() // NewLogger without arguments never errors
```

Comment:
I think we should avoid instantiaing new loggers with `EmptyContext`. A logger is meant to be a one-time instantiated thing to be passed and reused throughout. If some code isn't part of the request-response flow, it should just use `s.Log`/`app.Srv().Log`.

A logger is a heavy object and needs to be shutdown gracefully. Creating new ones every time creates additional overhead of ensuring every one of them is properly shut down.

Reply:
We are not initializing logger for every request. We are just adding the logger to the context along with the request_id. But maybe I misunderstood your question? :thinking:

Reply:
> I think we should avoid instantiaing new loggers with EmptyContext.

I found no way of getting a reference to the global logger instance. Is there a way to retrieve it?

> f some code isn't part of the request-response flow, it should just use s.Log/app.Srv().Log.

The problem is keeping the two cases apart. The methods are so entangled, that methods that are not part of requests call methods that are (== are called from methods that are). These expect a context with a logger now. To mitigate the problem (for now), I just create a new logger.

> A logger is a heavy object and needs to be shutdown gracefully. Creating new ones every time creates additional overhead of ensuring every one of them is properly shut down.

I didn't realize initializing a logger would create such an overhead.

> Thinking further, and probably it's a bit late to ask but why don't we use requestID as a parameter to s.Log? We would avoid initiating the logger for every request.

The logger has a trick up its sleeve to add info to a logger instance. So we are not creating a new logger but "pre-attaching" info to be logged to an existing one. From what I see it's pretty lightweight.
Just in a few cases I needed a logger instance without a clear path to retrieving one, so I added this `NewLogger()`.

Allowing the empty context to be initialized with an existing logger instance would be vetter, right?

Reply:
> The problem is keeping the two cases apart. The methods are so entangled, that methods that are not part of requests call methods that are (== are called from methods that are). 

IMO the distinction is clear: methods which are part of the request flow should call the logger from context. Methods which aren't part of the request flow, should call the logger in the `*Server` struct.

> Allowing the empty context to be initialized with an existing logger instance would be vetter, right?

Yes, that can work. But semantically it feels incorrect. From a user standpoint, calling logger methods on context should naturally mean that it has the request_id attached. But passing the original logger in the context doesn't serve much benefit other than using accessing logger via contexts everywhere.

Reply:
@noxer - let me know your thoughts on this

Reply:
The distinction is not that clear when we have methods that are call as part of a request but also as not part of a request. Which side do we put them on? IMHO those methods should take a context anyway. Then we run into the this issue, though. The other option would be to not use a context in this case but then we may miss log statements that are part of a request in the log because they are lacking the ID.